### PR TITLE
languages: add scala support

### PR DESCRIPTION
**Pull request by fusetim**
As a Scala lover, I would love to add the support for Scala, so I did and hope it would help.

##  Implementation

As Scala is build using the JVM, I only extended the existent Java class and made some adjustments to build Scala codes.
Furthermore, the current implementation use the same decompilation process (using `javap`) to find the main function, thus, Scala codes using `DelayedInit` or `App` traits just work out of the box.

I also added the requirements (for Debian) in the Travis-CI config.

### CODE REVIEW

File: .travis/install.sh [hunk 1]
```diff
@@ -24,7 +24,8 @@ sudo apt-get install -qq -y \
    php-cli \
    ruby \
    rustc \
-   swi-prolog-nox
+   swi-prolog-nox \
+   scala
```
Comment (alarsyo) [inline .travis/install.sh:28]: This dependency list looks like it was (approximately) sorted alphabetically, could you place the `scala` package above the `swi-prolog-nox` one?

**Review by alarsyo — COMMENTED** [review]
Hi, thanks for the contribution!

Just a quick modification for me.

I'm not familiar enough with `camisole` to judge the actual implementation, so I'll leave that to our in-house experts :nerd_face:

File: .travis/install.sh (hunk 1 above)
Comment (fusetim) [reply .travis/install.sh:28]: Oh, yes my mistake.

**Review by lportemo — APPROVED** [review]
Thanks for your contribution ;)

This looks good to me. All tests succeeded on archlinux as well.

File: README.rst [hunk 2]
```diff
@@ -29,7 +29,7 @@ Features
 
 - Built-in support for a wide variety of languages, including Ada, C, C#, C++,
   Haskell, Java, JavaScript, Lua, OCaml, Pascal, Perl, PHP, Python, Ruby,
-  Scheme…
+  Scheme, Scala…
```
Comment (Zopieux) [inline README.rst:32]: The list is sorted, please put Scheme last :-)

Comment (fusetim) [reply README.rst:32]: Dame it

File: camisole/languages/scala.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from camisole.languages.java import Java
+from camisole.models import Program
+
+class Scala(Java):
+    source_ext = '.scala'
+    compiled_ext = '.class'
+    compiler = Program('scalac',opts=['-encoding', 'UTF-8'],env={'LANG': 'en_US.UTF-8'}, version_opt='-version')
+    interpreter = Program('scala', version_opt='-version')
+    reference_source = r'object Main extends App {println("42");}'
```
Comment (Zopieux) [inline camisole/languages/scala.py:9]: Nit: aren't you missing spaces inside this `{ }` block?

File: camisole/languages/scala.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from camisole.languages.java import Java
+from camisole.models import Program
+
+class Scala(Java):
+    source_ext = '.scala'
+    compiled_ext = '.class'
+    compiler = Program('scalac',opts=['-encoding', 'UTF-8'],env={'LANG': 'en_US.UTF-8'}, version_opt='-version')
```
Comment (Zopieux) [inline camisole/languages/scala.py:7]: Please lint this code, whitespaces are borked.

**Review by Zopieux — CHANGES_REQUESTED** [review]

**Comment by Zopieux** [comment]
image

Scala is orders an order of magnitude worse on all metrics. Are scalac / scala programs really that terrible on both memory usage and cold wallclock runtime?

**Comment by fusetim** [comment]
Well, it's true that Scala cumulates the Java standard library and the Scala library, so it doesn't surprise me that it's worse than Java (in memory).
Also, contrary to Java, the Scala library is entirely contained in a jar that the JVM has to import at startup, which may explain this big discrepancy with Java (even if 2.5x surprises me a bit).
On the other hand, apart from this memory footprint due to the library, Scala and Java should remain quite close in terms of performance. Obviously on an example that displays only 42, the interest is to have the least overhead and having two standard libraries to load does not help.

It probably deserves more testing to get a clearer picture (and I'll try to provide you with some -- if I can recompile the JDK with the debug info).

**Review by Zopieux — APPROVED** [review]
Thanks!

**Comment by juli0z** [comment]
Migrated to https://gitlab.com/prologin/tech/services/camisole/-/merge_requests/53

**Pull request closed without merging**